File: javascript/src/CoordinateOperations.js
const UniformPrecisionCoordinateCompressor = require('./UniformPrecisionCoordinateCompressor');

/**
 * High-performance coordinate operations with batch processing capabilities.
 */
class CoordinateOperations {
    /**
     * Batch encodes multiple coordinate pairs for high-throughput scenarios.
     * @param {Array<{lat: number, lon: number}>} coordinates - Array of coordinate objects
     * @returns {string[]} Array of encoded Grid9 strings
     */
    static batchEncode(coordinates) {
        if (!Array.isArray(coordinates)) {
            throw new Error('Coordinates must be an array');
        }
        
        const results = new Array(coordinates.length);
        
        for (let i = 0; i < coordinates.length; i++) {
            const { lat, lon } = coordinates[i];
            if (typeof lat !== 'number' || typeof lon !== 'number') {
                throw new Error(`Invalid coordinate at index ${i}: lat and lon must be numbers`);
            }
            results[i] = UniformPrecisionCoordinateCompressor.encode(lat, lon);
        }
        
        return results;
    }

    /**
     * Batch decodes multiple encoded strings for high-throughput scenarios.
     * @param {string[]} encoded - Array of encoded Grid9 strings
     * @returns {Array<{lat: number, lon: number}>} Array of coordinate objects
     */
    static batchDecode(encoded) {
        if (!Array.isArray(encoded)) {
            throw new Error('Encoded must be an array');
        }
        
        const results = new Array(encoded.length);
        
        for (let i = 0; i < encoded.length; i++) {
            if (typeof encoded[i] !== 'string') {
                throw new Error(`Invalid encoded string at index ${i}: must be a string`);
            }
            const decoded = UniformPrecisionCoordinateCompressor.decode(encoded[i]);
            results[i] = { lat: decoded.latitude, lon: decoded.longitude };
        }
        
        return results;
    }

    /**
     * Finds all coordinates within a specified radius (in meters) of a center point.
     * Returns encoded strings of nearby coordinates.
     * @param {number} centerLat - Center latitude in degrees
     * @param {number} centerLon - Center longitude in degrees  
     * @param {number} radiusMeters - Search radius in meters
     * @param {number} [maxResults=100] - Maximum number of results to return
     * @returns {string[]} Array of encoded Grid9 strings within radius
     */
    static findNearby(centerLat, centerLon, radiusMeters, maxResults = 100) {
        if (typeof centerLat !== 'number' || typeof centerLon !== 'number') {
            throw new Error('Center coordinates must be numbers');
        }
        if (typeof radiusMeters !== 'number' || radiusMeters <= 0) {
            throw new Error('Radius must be a positive number');
        }
        if (typeof maxResults !== 'number' || maxResults <= 0) {
            throw new Error('Max results must be a positive number');
        }
        
        const results = [];
        const centerEncoded = UniformPrecisionCoordinateCompressor.encode(centerLat, centerLon);
        
        // Calculate approximate grid search bounds
        const latDelta = radiusMeters / 111320.0; // Rough conversion
        const lonDelta = radiusMeters / (111320.0 * Math.cos(centerLat * Math.PI / 180.0));
        
        const minLat = Math.max(-80, centerLat - latDelta);
        const maxLat = Math.min(80, centerLat + latDelta);
        const minLon = Math.max(-180, centerLon - lonDelta);
        const maxLon = Math.min(180, centerLon + lonDelta);
        
        // Grid search with 3-meter steps
        const latStep = 3.0 / 111320.0;
        const lonStep = 3.0 / 111320.0;
        
        for (let lat = minLat; lat <= maxLat && results.length < maxResults; lat += latStep) {
            for (let lon = minLon; lon <= maxLon && results.length < maxResults; lon += lonStep) {
                try {
                    const encoded = UniformPrecisionCoordinateCompressor.encode(lat, lon);
                    const distance = UniformPrecisionCoordinateCompressor.calculateDistance(centerEncoded, encoded);
                    
                    if (distance <= radiusMeters) {
                        results.push(encoded);
                    }
                } catch (error) {
                    // Skip invalid coordinates (out of bounds, etc.)
                    continue;
                }
            }
        }
        
        return results;
    }

    /**
     * Calculate the bounding box that contains all given coordinates
     * @param {Array<{lat: number, lon: number}>} coordinates - Array of coordinate objects
     * @returns {Object} Bounding box with minLat, maxLat, minLon, maxLon
     */
    static getBoundingBox(coordinates) {
        if (!Array.isArray(coordinates) || coordinates.length === 0) {
            throw new Error('Coordinates must be a non-empty array');
        }
        
        let minLat = Infinity;
        let maxLat = -Infinity;
        let minLon = Infinity;
        let maxLon = -Infinity;
        
        for (const coord of coordinates) {
            if (typeof coord.lat !== 'number' || typeof coord.lon !== 'number') {
                throw new Error('Each coordinate must have numeric lat and lon properties');
            }
            
            minLat = Math.min(minLat, coord.lat);
            maxLat = Math.max(maxLat, coord.lat);
            minLon = Math.min(minLon, coord.lon);
            maxLon = Math.max(maxLon, coord.lon);
        }
        
        return { minLat, maxLat, minLon, maxLon };
    }

    /**
     * Calculate the center point of a set of coordinates
     * @param {Array<{lat: number, lon: number}>} coordinates - Array of coordinate objects
     * @returns {Object} Center coordinate with lat and lon properties
     */
    static getCenterPoint(coordinates) {
        if (!Array.isArray(coordinates) || coordinates.length === 0) {
            throw new Error('Coordinates must be a non-empty array');
        }
        
        let totalLat = 0;
        let totalLon = 0;
        
        for (const coord of coordinates) {
            if (typeof coord.lat !== 'number' || typeof coord.lon !== 'number') {
                throw new Error('Each coordinate must have numeric lat and lon properties');
            }
            
            totalLat += coord.lat;
            totalLon += coord.lon;
        }
        
        return {
            lat: totalLat / coordinates.length,
            lon: totalLon / coordinates.length
        };
    }
}

module.exports = CoordinateOperations;

File: python/src/coordinate_operations.py
"""
High-performance coordinate operations with batch processing capabilities.
"""

import math
from typing import List, Tuple

from .uniform_precision_compressor import UniformPrecisionCoordinateCompressor


class CoordinateOperations:
    """High-performance coordinate operations with batch processing capabilities."""
    
    @staticmethod
    def batch_encode(coordinates: List[Tuple[float, float]]) -> List[str]:
        """
        Batch encodes multiple coordinate pairs for high-throughput scenarios.
        
        Args:
            coordinates: List of (latitude, longitude) tuples
            
        Returns:
            List of encoded strings
        """
        results = []
        
        for lat, lon in coordinates:
            results.append(UniformPrecisionCoordinateCompressor.encode(lat, lon))
        
        return results
    
    @staticmethod
    def batch_decode(encoded: List[str]) -> List[Tuple[float, float]]:
        """
        Batch decodes multiple encoded strings for high-throughput scenarios.
        
        Args:
            encoded: List of encoded strings
            
        Returns:
            List of (latitude, longitude) tuples
        """
        results = []
        
        for enc_str in encoded:
            results.append(UniformPrecisionCoordinateCompressor.decode(enc_str))
        
        return results
    
    @staticmethod
    def find_nearby(center_lat: float, center_lon: float, radius_meters: float, max_results: int = 100) -> List[str]:
        """
        Finds all coordinates within a specified radius (in meters) of a center point.
        Returns encoded strings of nearby coordinates.
        
        Args:
            center_lat: Center latitude in degrees
            center_lon: Center longitude in degrees
            radius_meters: Search radius in meters
            max_results: Maximum number of results to return (default: 100)
            
        Returns:
            List of encoded coordinate strings within the radius
        """
        results = []
        center_encoded = UniformPrecisionCoordinateCompressor.encode(center_lat, center_lon)
        
        # Calculate approximate grid search bounds
        lat_delta = radius_meters / 111320.0  # Rough conversion
        lon_delta = radius_meters / (111320.0 * math.cos(math.radians(center_lat)))
        
        min_lat = max(-80, center_lat - lat_delta)
        max_lat = min(80, center_lat + lat_delta)
        min_lon = max(-180, center_lon - lon_delta)
        max_lon = min(180, center_lon + lon_delta)
        
        # Grid search with 3-meter steps
        lat_step = 3.0 / 111320.0
        lon_step = 3.0 / 111320.0
        
        lat = min_lat
        while lat <= max_lat and len(results) < max_results:
            lon = min_lon
            while lon <= max_lon and len(results) < max_results:
                try:
                    encoded = UniformPrecisionCoordinateCompressor.encode(lat, lon)
                    distance = UniformPrecisionCoordinateCompressor.calculate_distance(center_encoded, encoded)
                    
                    if distance <= radius_meters:
                        results.append(encoded)
                except ValueError:
                    # Skip invalid coordinates (out of bounds, etc.)
                    pass
                
                lon += lon_step
            lat += lat_step
        
        return results